File: apps/apple/core/Sources/LorvexWorkflow/TaskDeferral.swift
import Foundation
import GRDB
import LorvexDomain
import LorvexStore

/// Canonical task-deferral SQL mutations.
///
/// `defer_task` increments `defer_count`, stamps `last_deferred_at`,
/// optionally writes `planned_date`, `ai_notes`, and `last_defer_reason`
/// in a single LWW-gated UPDATE. When the planned date changes, pending
/// reminders (not dismissed, not cancelled, not yet fired) for the task
/// are shifted by the same calendar-day delta, so a deferred task carries
/// its pending reminders with it. The reminder HLC stamps come from the caller via
/// the `nextReminderVersion` closure because each runtime owns its HLC
/// session.
///
/// Each entry returns `updated == false` when the LWW gate or status
/// filter rejects the write — callers re-stamp the HLC and retry.
public enum TaskDeferral {
  public struct DeferralPatch: Sendable {
    public var plannedDate: String?
    public var aiNotes: String?
    public var lastDeferReason: String?

    public init(
      plannedDate: String? = nil,
      aiNotes: String? = nil,
      lastDeferReason: String? = nil
    ) {
      self.plannedDate = plannedDate
      self.aiNotes = aiNotes
      self.lastDeferReason = lastDeferReason
    }
  }

  public struct DeferralResult: Sendable, Equatable {
    public var updated: Bool
    public var shiftedReminderIds: [String]

    public init(updated: Bool = false, shiftedReminderIds: [String] = []) {
      self.updated = updated
      self.shiftedReminderIds = shiftedReminderIds
    }
  }

  /// Multi-field snapshot of the pre-mutation deferral state that
  /// `restoreTaskDeferral` stamps back onto the row.
  public struct DeferralSnapshot: Sendable {
    public var plannedDate: String?
    public var deferCount: Int64
    public var lastDeferredAt: String?
    public var lastDeferReason: String?

    public init(
      plannedDate: String? = nil,
      deferCount: Int64 = 0,
      lastDeferredAt: String? = nil,
      lastDeferReason: String? = nil
    ) {
      self.plannedDate = plannedDate
      self.deferCount = deferCount
      self.lastDeferredAt = lastDeferredAt
      self.lastDeferReason = lastDeferReason
    }
  }

  /// Apply a single task deferral atomically inside a SAVEPOINT named
  /// `task_deferral`. Returns `updated == false` when the LWW gate
  /// rejects the stamp or the status filter excludes the row
  /// (`completed`, `cancelled`, or non-existent).
  public static func deferTask(
    _ db: Database,
    taskId: TaskId,
    patch: DeferralPatch,
    version: String,
    now: String,
    nextReminderVersion: () throws -> String
  ) throws -> DeferralResult {
    // `last_defer_reason` must be one of the canonical `DeferReason` categories.
    // Validating here — the single core write path every surface funnels through
    // (MCP, batch, intents) — means no caller can reach the raw
    // `CHECK (last_defer_reason IN (...))` and surface an opaque
    // `SQLITE_CONSTRAINT`; a bad category fails with a typed validation error.
    if let reason = patch.lastDeferReason, DeferReason.parse(reason) == nil {
      throw StoreError.validation(
        "last_defer_reason \"\(reason)\" is not a recognized defer reason "
          + "(expected one of: \(DeferReasonName.allDeferReasons.joined(separator: ", ")))")
    }
    try db.execute(sql: "SAVEPOINT task_deferral")
    do {
      let result = try deferTaskInner(
        db, taskId: taskId, patch: patch, version: version, now: now,
        nextReminderVersion: nextReminderVersion)
      try db.execute(sql: "RELEASE SAVEPOINT task_deferral")
      return result
    } catch {
      try? db.execute(sql: "ROLLBACK TO SAVEPOINT task_deferral")
      try? db.execute(sql: "RELEASE SAVEPOINT task_deferral")
      throw error
    }
  }

  private static func deferTaskInner(
    _ db: Database,
    taskId: TaskId,
    patch: DeferralPatch,
    version: String,
    now: String,
    nextReminderVersion: () throws -> String
  ) throws -> DeferralResult {
    let reminderShift: ReminderShiftContext?
    if let newPlanned = patch.plannedDate {
      reminderShift = try loadReminderShiftContext(
        db, taskId: taskId, newPlannedDate: newPlanned)
    } else {
      reminderShift = nil
    }

    var setClauses: [String] = [
      "defer_count = MIN(defer_count + 1, 9223372036854775807)"
    ]
    var args: [(any DatabaseValueConvertible)?] = []

    if let date = patch.plannedDate {
      args.append(date)
      setClauses.append("planned_date = ?\(args.count)")
    }
    if let notes = patch.aiNotes {
      args.append(notes)
      setClauses.append("ai_notes = ?\(args.count)")
    }
    if let reason = patch.lastDeferReason {
      args.append(reason)
      setClauses.append("last_defer_reason = ?\(args.count)")
    }

    args.append(now)
    let nowIdx = args.count
    setClauses.append("last_deferred_at = ?\(nowIdx)")
    setClauses.append("updated_at = ?\(nowIdx)")

    args.append(version)
    let versionIdx = args.count
    setClauses.append("schedule_version = ?\(versionIdx)")
    if patch.aiNotes != nil {
      setClauses.append("content_version = ?\(versionIdx)")
    }
    setClauses.append("version = ?\(versionIdx)")

    args.append(taskId.rawValue)
    let idIdx = args.count

    let sql =
      "UPDATE tasks SET \(setClauses.joined(separator: ", ")) "
      + "WHERE id = ?\(idIdx) AND status NOT IN ('completed', 'cancelled') "
      + "AND ?\(versionIdx) > version"

    try db.execute(sql: sql, arguments: StatementArguments(args))
    if db.changesCount == 0 {
      return DeferralResult()
    }

    let shiftedIds: [String]
    if let ctx = reminderShift {
      shiftedIds = try shiftPendingRemindersToNewPlannedDate(
        db, taskId: taskId, context: ctx, now: now,
        nextReminderVersion: nextReminderVersion)
    } else {
      shiftedIds = []
    }

    return DeferralResult(updated: true, shiftedReminderIds: shiftedIds)
  }

  /// Reset task deferral state: clear `planned_date`, `last_deferred_at`,
  /// `last_defer_reason`, and reset `defer_count` to 0. Returns `false`
  /// when the LWW gate rejects or the row is terminal / missing.
  public static func resetTaskDeferral(
    _ db: Database,
    taskId: TaskId,
    version: String,
    now: String
  ) throws -> Bool {
    try db.execute(
      sql:
        "UPDATE tasks SET "
        + "planned_date = NULL, "
        + "last_deferred_at = NULL, "
        + "last_defer_reason = NULL, "
        + "defer_count = 0, "
        + "schedule_version = ?, "
        + "version = ?, "
        + "updated_at = ? "
        + "WHERE id = ? AND status NOT IN ('completed', 'cancelled') "
        + "AND ? > version",
      arguments: [version, version, now, taskId.rawValue, version])
    return db.changesCount > 0
  }

  /// Restore the exact pre-defer deferral state captured in `snapshot`.
  /// Used by the single-action "Undo" toast path. LWW-gated; returns
  /// `false` when the gate rejects the stamp.
  public static func restoreTaskDeferral(
    _ db: Database,
    taskId: TaskId,
    snapshot: DeferralSnapshot,
    version: String,
    now: String
  ) throws -> Bool {
    try db.execute(
      sql:
        "UPDATE tasks SET "
        + "planned_date = ?, "
        + "defer_count = ?, "
        + "last_deferred_at = ?, "
        + "last_defer_reason = ?, "
        + "schedule_version = ?, "
        + "version = ?, "
        + "updated_at = ? "
        + "WHERE id = ? AND status NOT IN ('completed', 'cancelled') "
        + "AND ? > version",
      arguments: [
        snapshot.plannedDate, snapshot.deferCount, snapshot.lastDeferredAt,
        snapshot.lastDeferReason, version, version, now, taskId.rawValue, version,
      ])
    return db.changesCount > 0
  }

  // MARK: - Reminder shift

  private struct ReminderShiftContext {
    let oldReferenceDate: String
    let newPlannedDate: String
  }

  private static func loadReminderShiftContext(
    _ db: Database, taskId: TaskId, newPlannedDate: String
  ) throws -> ReminderShiftContext? {
    let row = try Row.fetchOne(
      db,
      sql: "SELECT COALESCE(planned_date, due_date) FROM tasks WHERE id = ?",
      arguments: [taskId.rawValue])
    guard let row else { return nil }
    let old: String? = row[0]
    guard let old else { return nil }
    return ReminderShiftContext(
      oldReferenceDate: old, newPlannedDate: newPlannedDate)
  }

  /// Walk all future pending reminders for the task and rewrite each
  /// `reminder_at` by the calendar-day delta between the old
  /// planned-or-due reference date and the new planned date. Each rewrite
  /// stamps a fresh HLC version pulled from the caller.
  ///
  /// Both reference dates come from validated write paths, so a parse
  /// failure implies on-disk corruption — surfaced as
  /// ``StoreError/validation`` so the deferral fails loudly rather than
  /// silently leaving stale reminders.
  private static func shiftPendingRemindersToNewPlannedDate(
    _ db: Database,
    taskId: TaskId,
    context: ReminderShiftContext,
    now: String,
    nextReminderVersion: () throws -> String
  ) throws -> [String] {
    let oldDate = try parseYMD(
      context.oldReferenceDate, label: "old_reference_date")
    let newDate = try parseYMD(
      context.newPlannedDate, label: "new_planned_date")
    let deltaDays = IsoDate.dayNumber(newDate) - IsoDate.dayNumber(oldDate)
    if deltaDays == 0 { return [] }

    // Shift reminders by calendar days in the user's anchored timezone so they
    // keep their local wall-clock time across a DST boundary.
    let reminderTimezone =
      TimeZone(identifier: try WorkflowTimezone.anchoredTimezoneName(db))
      ?? TimeZone(secondsFromGMT: 0)!

    let rows = try Row.fetchAll(
      db,
      sql:
        "SELECT id, reminder_at FROM task_reminders "
        + "WHERE task_id = ? AND dismissed_at IS NULL AND cancelled_at IS NULL "
        + "AND reminder_at > ?",
      arguments: [taskId.rawValue, now])

    var shifted: [String] = []
    for row in rows {
      let reminderId: String = row[0]
      let reminderAt: String = row[1]
      let shiftedAt = try shiftRFC3339(reminderAt, byDays: deltaDays,
        timezone: reminderTimezone, reminderId: reminderId)
      let reminderVersion = try nextReminderVersion()
      try db.execute(
        sql:
          "UPDATE task_reminders SET reminder_at = ?, version = ? "
          + "WHERE id = ? AND ? > version",
        arguments: [shiftedAt, reminderVersion, reminderId, reminderVersion])
      if db.changesCount > 0 {
        shifted.append(reminderId)
      }
    }
    return shifted
  }

  // MARK: - Date helpers

  /// Parse a canonical `YYYY-MM-DD` string into a validated calendar date via
  /// the shared ``IsoDate/parse(_:)``. Throws ``StoreError/validation`` on
  /// malformed input.
  private static func parseYMD(_ s: String, label: String) throws -> IsoDate.YMD {
    guard let ymd = IsoDate.parse(s) else {
      throw StoreError.validation(
        "shift_pending_reminders: corrupt \(label) \"\(s)\""
      )
    }
    return ymd
  }

  /// Shift an RFC3339 timestamp by `days` calendar days in `timezone`,
  /// preserving the fractional-second / `Z` sync-timestamp formatting.
  ///
  /// Advancing by *calendar* days in the user's timezone (rather than a fixed
  /// `days * 86400` seconds) keeps the reminder's local wall-clock time across a
  /// DST transition — otherwise a 09:00 reminder deferred across spring-forward
  /// would re-fire at 10:00.
  private static func shiftRFC3339(
    _ timestamp: String, byDays days: Int, timezone: TimeZone, reminderId: String
  ) throws -> String {
    guard let parsed = ReminderAnchor.parseRfc3339ToDate(timestamp) else {
      throw StoreError.validation(
        "shift_pending_reminders: corrupt stored reminder_at \"\(timestamp)\" for reminder \(reminderId)"
      )
    }
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = timezone
    guard let shifted = calendar.date(byAdding: .day, value: days, to: parsed) else {
      throw StoreError.validation(
        "shift_pending_reminders: could not shift reminder_at \"\(timestamp)\" by \(days) days for reminder \(reminderId)"
      )
    }
    return SyncTimestampFormat.formatSyncTimestamp(shifted)
  }
}

File: apps/tauri/lorvex-workflow/src/task_deferral.rs
//! Shared task deferral operation.
//!
//! Both the MCP server and Tauri app delegate the core SQL mutation here.
//! Adapter-specific concerns (changelog, sync enqueue) remain in each boundary,
//! but `ai_notes` is folded into the atomic UPDATE via `TaskDeferralPatch`.

use lorvex_domain::TaskId;
use rusqlite::{Connection, OptionalExtension};

use lorvex_store::StoreError;

/// Parameters for a task deferral.
#[derive(Debug, Clone, Default)]
pub struct TaskDeferralPatch<'a> {
    /// New planned date. Some = set, None = leave unchanged.
    pub planned_date: Option<&'a str>,
    /// New ai_notes value. Some = overwrite, None = leave unchanged.
    pub ai_notes: Option<&'a str>,
    /// Structured defer reason. Some = set, None = leave unchanged.
    pub last_defer_reason: Option<&'a str>,
}

/// Result of applying a task deferral.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct TaskDeferralResult {
    pub updated: bool,
    pub shifted_reminder_ids: Vec<String>,
}

#[derive(Debug, Clone)]
struct ReminderShiftContext {
    old_reference_date: String,
    new_planned_date: String,
}

/// Apply a single task deferral atomically.
///
/// Updates `planned_date` (if provided), `ai_notes` (if provided),
/// `last_defer_reason` (if provided), increments `defer_count`, stamps `last_deferred_at`, `version`,
/// and `updated_at` in a single UPDATE.
///
/// If the planned date changes, pending reminders anchored to the task's
/// previous planned/due date move by the same calendar-day delta. The caller
/// supplies reminder HLC stamps because each runtime owns its HLC state, while
/// the store owns the common SQL semantics.
pub fn defer_task<E>(
    conn: &Connection,
    task_id: &TaskId,
    patch: &TaskDeferralPatch<'_>,
    version: &str,
    now: &str,
    mut next_reminder_version: impl FnMut() -> Result<String, E>,
) -> Result<TaskDeferralResult, E>
where
    E: From<rusqlite::Error> + From<StoreError> + From<String> + std::fmt::Display,
{
    lorvex_store::transaction::with_savepoint(conn, "task_deferral", |conn| {
        defer_task_inner(
            conn,
            task_id,
            patch,
            version,
            now,
            &mut next_reminder_version,
        )
    })
}

fn defer_task_inner<E>(
    conn: &Connection,
    task_id: &TaskId,
    patch: &TaskDeferralPatch<'_>,
    version: &str,
    now: &str,
    next_reminder_version: &mut impl FnMut() -> Result<String, E>,
) -> Result<TaskDeferralResult, E>
where
    E: From<StoreError>,
{
    let reminder_shift = match patch.planned_date {
        Some(new_planned_date) => load_reminder_shift_context(conn, task_id, new_planned_date)?,
        None => None,
    };

    // Build dynamic SET clauses
    let mut set_clauses =
        vec!["defer_count = MIN(defer_count + 1, 9223372036854775807)".to_string()];
    let mut params: Vec<&dyn rusqlite::types::ToSql> = Vec::new();

    if let Some(ref date) = patch.planned_date {
        params.push(date);
        set_clauses.push(format!("planned_date = ?{}", params.len()));
    }
    if let Some(ref notes) = patch.ai_notes {
        params.push(notes);
        set_clauses.push(format!("ai_notes = ?{}", params.len()));
    }
    if let Some(ref reason) = patch.last_defer_reason {
        params.push(reason);
        set_clauses.push(format!("last_defer_reason = ?{}", params.len()));
    }

    // Always set these
    params.push(&now);
    let now_idx = params.len();
    set_clauses.push(format!("last_deferred_at = ?{now_idx}"));
    set_clauses.push(format!("updated_at = ?{now_idx}"));

    params.push(&version);
    let version_idx = params.len();
    set_clauses.push(format!("version = ?{version_idx}"));

    params.push(&task_id);
    let id_idx = params.len();

    // gate the UPDATE on `?version_idx > version` so a
    // stale local stamp loses to an in-flight peer write that already
    // landed a newer version. Returns `Ok(false)` so the boundary
    // layer can re-stamp HLC and retry — same shape as the existing
    // "task already terminal" no-op return.
    let sql = format!(
        "UPDATE tasks SET {} \
         WHERE id = ?{id_idx} AND status NOT IN ('completed', 'cancelled') \
         AND ?{version_idx} > version",
        set_clauses.join(", ")
    );

    let changes = conn
        .execute(&sql, params.as_slice())
        .map_err(StoreError::from)?;
    if changes == 0 {
        return Ok(TaskDeferralResult::default());
    }

    let shifted_reminder_ids = match reminder_shift {
        Some(context) => shift_pending_reminders_to_new_planned_date(
            conn,
            task_id,
            &context,
            now,
            next_reminder_version,
        )?,
        None => Vec::new(),
    };

    Ok(TaskDeferralResult {
        updated: true,
        shifted_reminder_ids,
    })
}

fn load_reminder_shift_context(
    conn: &Connection,
    task_id: &TaskId,
    new_planned_date: &str,
) -> Result<Option<ReminderShiftContext>, StoreError> {
    let old_reference_date: Option<String> = conn
        .query_row(
            "SELECT COALESCE(planned_date, due_date) FROM tasks WHERE id = ?1",
            rusqlite::params![task_id],
            |row| row.get(0),
        )
        .optional()?
        .flatten();
    Ok(
        old_reference_date.map(|old_reference_date| ReminderShiftContext {
            old_reference_date,
            new_planned_date: new_planned_date.to_string(),
        }),
    )
}

fn shift_pending_reminders_to_new_planned_date<E>(
    conn: &Connection,
    task_id: &TaskId,
    context: &ReminderShiftContext,
    now: &str,
    next_reminder_version: &mut impl FnMut() -> Result<String, E>,
) -> Result<Vec<String>, E>
where
    E: From<StoreError>,
{
    // Audit (silent-failure-hunter):
    // here returned empty / continued silently, so a corrupt stored
    // date string would defer the task without shifting reminders and
    // the user got no signal. Both reference dates came from validated
    // write paths, so a parse failure implies on-disk corruption —
    // surface as `StoreError::Validation` so the deferral fails loudly
    // rather than silently leaving stale reminders.
    let old_date = chrono::NaiveDate::parse_from_str(&context.old_reference_date, "%Y-%m-%d")
        .map_err(|e| {
            StoreError::Validation(format!(
                "shift_pending_reminders: corrupt old_reference_date {:?}: {e}",
                context.old_reference_date
            ))
        })?;
    let new_date = chrono::NaiveDate::parse_from_str(&context.new_planned_date, "%Y-%m-%d")
        .map_err(|e| {
            StoreError::Validation(format!(
                "shift_pending_reminders: corrupt new_planned_date {:?}: {e}",
                context.new_planned_date
            ))
        })?;
    let delta_days = (new_date - old_date).num_days();
    if delta_days == 0 {
        return Ok(Vec::new());
    }

    let pending_reminders = {
        let mut stmt = conn
            .prepare_cached(
                "SELECT id, reminder_at FROM task_reminders \
                 WHERE task_id = ?1 AND dismissed_at IS NULL AND cancelled_at IS NULL \
                   AND reminder_at > ?2",
            )
            .map_err(StoreError::from)?;
        let rows = stmt
            .query_map(rusqlite::params![task_id, now], |row| {
                Ok((row.get::<_, String>(0)?, row.get::<_, String>(1)?))
            })
            .map_err(StoreError::from)?
            .collect::<Result<Vec<_>, _>>()
            .map_err(StoreError::from)?;
        rows
    };

    let mut shifted_ids = Vec::new();
    for (reminder_id, reminder_at_str) in pending_reminders {
        let parsed = chrono::DateTime::parse_from_rfc3339(&reminder_at_str).map_err(|e| {
            StoreError::Validation(format!(
                "shift_pending_reminders: corrupt stored reminder_at {reminder_at_str:?} for reminder {reminder_id}: {e}"
            ))
        })?;
        let shifted = parsed + chrono::Duration::days(delta_days);
        let new_reminder_at =
            lorvex_domain::format_sync_timestamp(shifted.with_timezone(&chrono::Utc));
        let reminder_version = next_reminder_version()?;
        let changes = conn
            .execute(
                "UPDATE task_reminders SET reminder_at = ?1, version = ?2 \
                 WHERE id = ?3 AND ?2 > version",
                rusqlite::params![new_reminder_at, reminder_version, reminder_id],
            )
            .map_err(StoreError::from)?;
        if changes > 0 {
            shifted_ids.push(reminder_id);
        }
    }
    Ok(shifted_ids)
}

/// Reset task deferral state: clear planned_date, last_deferred_at,
/// last_defer_reason, and reset defer_count to 0.
///
/// gated by `?1 > version`. A stale local stamp
/// returns `Ok(false)` so the caller can re-stamp HLC and retry.
pub fn reset_task_deferral(
    conn: &Connection,
    task_id: &TaskId,
    version: &str,
    now: &str,
) -> Result<bool, StoreError> {
    let changes = conn
        .prepare_cached(
            "UPDATE tasks SET \
             planned_date = NULL, \
             last_deferred_at = NULL, \
             last_defer_reason = NULL, \
             defer_count = 0, \
             version = ?1, \
             updated_at = ?2 \
             WHERE id = ?3 AND status NOT IN ('completed', 'cancelled') \
             AND ?1 > version",
        )?
        .execute(rusqlite::params![version, now, task_id])?;
    Ok(changes > 0)
}

#[derive(Debug, Clone)]
pub struct TaskDeferralSnapshot<'a> {
    pub planned_date: Option<&'a str>,
    pub defer_count: i64,
    pub last_deferred_at: Option<&'a str>,
    pub last_defer_reason: Option<&'a str>,
}

/// Restore the exact pre-defer deferral state captured in `snapshot`.
/// Used by the single-action "Undo" toast path.
///
/// gated by `?5 > version`. A stale local stamp
/// returns `Ok(false)` so the caller can re-stamp HLC and retry
/// instead of clobbering a freshly-applied peer write.
pub fn restore_task_deferral(
    conn: &Connection,
    task_id: &TaskId,
    snapshot: &TaskDeferralSnapshot<'_>,
    version: &str,
    now: &str,
) -> Result<bool, StoreError> {
    let changes = conn
        .prepare_cached(
            "UPDATE tasks SET \
             planned_date = ?1, \
             defer_count = ?2, \
             last_deferred_at = ?3, \
             last_defer_reason = ?4, \
             version = ?5, \
             updated_at = ?6 \
             WHERE id = ?7 AND status NOT IN ('completed', 'cancelled') \
             AND ?5 > version",
        )?
        .execute(rusqlite::params![
            snapshot.planned_date,
            snapshot.defer_count,
            snapshot.last_deferred_at,
            snapshot.last_defer_reason,
            version,
            now,
            task_id,
        ])?;
    Ok(changes > 0)
}

#[cfg(test)]
mod tests;
